src/osd/scrubber/scrub_machine.h

```diff
@@ -776,16 +839,75 @@ struct ReplicaActive : sc::state<
       const ReplicaReserveReq&,
       bool async_request);
 
+  /**
+   * the queued reservation request was granted by the async reserver.
+   * Notify the Primary.
+   * Returns 'false' if the reservation is not the last one to be received
+   * by this replica.
+   */
+  bool granted_by_reserver(const AsyncScrubResData& resevation);
+
   /// handle a 'release' from a primary
   void on_release(const ReplicaRelease& ev);
 
-  /// cancel the reserver request.
-  /// The 'failure' re 'log_failure' is logged if we are not reserved to
-  /// begin with.
+  /**
+   * cancel the reserver request.
+   * The 'failure' re 'log_failure' is logged if we are not reserved to
+   * begin with.
+   */
   void clear_reservation_by_remote_primary(bool log_failure);
 
+  /**
+   * cancel_ungranted_request - cancel the reservation request if it was not
+   * already granted by the async reserver.
+   */
+  void cancel_ungranted_request();
+
   using reactions = mpl::list<sc::transition<IntervalChanged, NotActive>>;
 
+  // clang-format off
+  // remote reservation machinery
+  struct RtReservationCB : public Context {
+    PGRef pg;
+    AsyncScrubResData res_data;
+    bool canceled{false};
+
+    explicit RtReservationCB(PGRef pg, AsyncScrubResData request_details)
+	: pg{pg}
+	, res_data{request_details}
+    {}
+
+    void finish(int) override {
+      pg->lock();
+      if (!canceled)
+	pg->m_scrubber->send_granted_by_reserver(res_data);
+      pg->unlock();
+    }
+
+    void cancel() {
+      ceph_assert(pg->is_locked());
+      ceph_assert(!canceled);
+      canceled = true;
+    }
+  };
+  // clang-format on
+
+  Context* reserver_callback{nullptr};
+  MOSDScrubReserve::reservation_nonce_t last_request_nonce{0};
+
+  // clang-format off
+  struct NullOpCB : public Context {
+    PGRef pg;
+    bool canceled{false};
+    explicit NullOpCB(PGRef pg) : pg{pg} {}
+    void finish(int) override {}
+    void cancel() {
+      ceph_assert(pg->is_locked());
+      canceled = true;
+    }
+  };
+  // clang-format on
+
  private:
   bool reserved_by_my_primary{false};
```

Comment:
You didn't add this in this PR, but this is a really confusing name.  It's not possible to be reserved by anything other than the primary of the current interval.  The name could be interpreted to indicate either:
1. a reservation is active
2. that the active reservation is specifically for the primary

I'd really like the name changed to reservation_granted or something with a comment explaining that it is set to true iff we have sent a grant confirmation to the primary and the reservation has not yet been canceled either by the primary or by an interval change.

Reply:
That would give us two state variables that fully express the current reservation state (pending_reservation_nonce and reservation_granted) and 3 possible states:
1. pending_reservation_nonce == 0 && !reservation_granted -- no reservation is requested or pending
2. pending_reservation_nonce != 0 && !reservation_granted -- we have a pending cb in the AsyncReserver for a request with nonce pending_reservation_nonce.
3. pending_reservation_nonce == 0 && reservation_granted -- we have sent a response to the primary granting the reservation

The syncronous/compatibility case bypasses 2.

Invariant: assert(!((pending_reservation_nonce != 0) && reservation_granted));

Whether you choose to adopt the above arrangement or not, I'd like to see a comment spelling out precisely what member variable combinations are possible.